File: cpp/837_新21点_medium.cpp
class Solution {
public:
    double new21Game(int N, int K, int W) {
        if(K==0)
            return 1.0;
        vector<double> dp(K+W, 0.0);
        for(int i=K; i<=min(N, (K+W-1)); i++)
            dp[i] = 1.0;
        dp[K-1] = (min(N, K+W-1)-K+1)*1.0/W;
        for(int i=K-2; i>-1; i--)
            dp[i] = dp[i+1] - (dp[i+W+1] - dp[i+1])/W;
        return dp[0];
    }
};

File: python/837_新21点_medium.py
class Solution:
    def new21Game(self, N: int, K: int, W: int) -> float:
        if K==0:
            return 1.0
        dp = [0.0]*(K+W)
        for i in range(K, min(N, K+W-1)+1):
            dp[i] = 1.0
        dp[K-1] = (min(N, K+W-1) - K + 1)*1.0/W
        for i in range(K-2, -1, -1):
            dp[i] = dp[i+1] - (dp[i+W+1] - dp[i+1])/W
        return dp[0]